pepper-apis/src/main/java/org/broadinstitute/ddp/service/KitCheckService.java

```diff
@@ -63,123 +62,162 @@ public KitCheckService(int batchSize) {
      * initial kit).
      *
-     * @param apisHandle the database handle
      * @return mapping of study to number of participants queued in this run
      */
-    public KitCheckResult scheduleNextKits(Handle apisHandle) {
-        var kitConfigs = apisHandle.attach(KitConfigurationDao.class)
-                .kitConfigurationFactory()
-                .stream()
-                .filter(config -> config.getSchedule() != null)
-                .filter(config -> config.getNumKits() >= 0)
-                .collect(Collectors.toList());
+    public KitCheckResult scheduleNextKits() {
+        Queue<KitConfiguration> kitConfigs = withAPIsTxn(handle -> new ArrayDeque<>(
+                handle.attach(KitConfigurationDao.class)
+                        .kitConfigurationFactory()
+                        .stream()
+                        .filter(config -> config.getSchedule() != null)
+                        .filter(config -> config.getNumKits() >= 0)
+                        .collect(Collectors.toList())));
 
-        var kitScheduleDao = apisHandle.attach(KitScheduleDao.class);
         var kitCheckResult = new KitCheckResult();
 
-        for (var kitConfig : kitConfigs) {
-            List<PendingScheduleRecord> batch = new ArrayList<>();
-            kitScheduleDao.findPendingScheduleRecords(kitConfig.getId()).forEach(pending -> {
-                batch.add(pending);
-                if (batch.size() >= batchSize) {
-                    scheduleNextKitsForBatch(apisHandle, kitCheckResult, kitConfig, batch);
-                    batch.clear();
+        while (!kitConfigs.isEmpty()) {
+            KitConfiguration kitConfig = kitConfigs.remove();
+            long kitConfigId = kitConfig.getId();
+            String studyGuid = kitConfig.getStudyGuid();
+            LOG.info("Checking kit schedule records for study {} and kit configuration {}", studyGuid, kitConfigId);
+
+            int fetched = 0;
+            while (true) {
+                int offset = fetched;
+                Queue<PendingScheduleRecord> batch = withAPIsTxn(handle -> new ArrayDeque<>(
+                        handle.attach(KitScheduleDao.class)
+                                .findPendingScheduleRecords(kitConfig.getId(), offset, batchSize)));
+                int fetchedSize = batch.size();
+                if (fetchedSize == 0) {
+                    break;
                 }
-            });
-            if (!batch.isEmpty()) {
-                scheduleNextKitsForBatch(apisHandle, kitCheckResult, kitConfig, batch);
+                while (!batch.isEmpty()) {
+                    PendingScheduleRecord record = batch.remove();
+                    try {
+                        scheduleNextKitForParticipant(kitCheckResult, kitConfig, record);
+                    } catch (Exception e) {
+                        LOG.error("Error while checking kit schedule record for participant {}, continuing", record.getUserGuid(), e);
+                    }
+                }
```

Comment:
Might be possible to package this type of iteration of set over multiple queries and multiple transactions into a utility. No need to do it now, but if this becomes a recurring pattern...

Reply:
Yup, it's certainly something I have though about since I've written it a few times. But haven't found the right pattern to expose this as a utility yet.

Reply:
I have a vague idea of exposing this kind of pattern as an `Iterator`, but I'll flush that out in a separate refactor later.